### PR TITLE
feat: support SDK auth in AzureAD RemoteWrite

### PR DESCRIPTION
**Pull request by slashpai**
feat: support SDK auth in AzureAD RemoteWrite

Related-to #6584

## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
feat: support SDK auth in AzureAD RemoteWrite
```

### CODE REVIEW

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 1]
```diff
@@ -1419,6 +1426,15 @@ type ManagedIdentity struct {
 	ClientID string `json:"clientId"`
 }
 
+// AzureSDK is used to store azure SDK config values.
+type AzureSDK struct {
+	// `tenantId` is the tenant ID of the azure active directory application that is being used to authenticate.
+	// +required
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern:=^[0-9a-zA-Z-.]+$
+	TenantID string `json:"tenantId,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:1435]: Looking at the Prometheus code, tenant id is optional.

Comment (slashpai) [reply pkg/apis/monitoring/v1/prometheus_types.go]: ah rigjt, my bad corrected it now

**Comment by slashpai** [comment]
@simonpasquier addressed comments PTAL

**Review by simonpasquier — COMMENTED** [review]
we need to update `ValidateRemoteWriteSpec` to ensure that only one auth method is defined.

File: pkg/prometheus/promcfg.go [hunk 2]
```diff
@@ -2124,6 +2124,12 @@ func (cg *ConfigGenerator) generateRemoteWriteConfig(
 				})
 			}
 
+			if spec.AzureAD.SDK != nil {
+				azureAd = cg.WithMinimumVersion("2.52.0").AppendMapItem(azureAd, "sdk", yaml.MapSlice{
+					{Key: "tenant_id", Value: spec.AzureAD.SDK.TenantID},
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2129]:
```suggestion
					{Key: "tenant_id", Value: ptr.Deref(spec.AzureAD.SDK.TenantID, "")},
```

File: pkg/prometheus/operator.go [hunk 3]
```diff
@@ -127,14 +127,22 @@ func ValidateRemoteWriteSpec(spec monitoringv1.RemoteWriteSpec) error {
 	}
 
 	if spec.AzureAD != nil {
-		if spec.AzureAD.ManagedIdentity == nil && spec.AzureAD.OAuth == nil {
-			return fmt.Errorf("must provide Azure Managed Identity or Azure OAuth in the Azure AD config")
+		if spec.AzureAD.ManagedIdentity == nil && spec.AzureAD.OAuth == nil && spec.AzureAD.SDK == nil {
```
Comment (simonpasquier) [inline pkg/prometheus/operator.go:130]: There's still a risk that users pick up the SDK option with an older Prometheus version in which the generated config will fail. Ideally the operator should fail here.

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -2124,6 +2124,12 @@ func (cg *ConfigGenerator) generateRemoteWriteConfig(
 				})
 			}
 
+			if spec.AzureAD.SDK != nil {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2127]: (nit) we could use a switch/case to make it more readable/robust.

**Review by simonpasquier — APPROVED** [review]
I have one late comment but we can have the improvement in a follow-up PR.

**Comment by slashpai** [comment]
I can address comments in follow up PR I had enabled auto-merge :)

File: pkg/prometheus/operator.go (hunk 3 above)
Comment (slashpai) [reply pkg/prometheus/operator.go:130]: should we have this validated similar to relabel configs is resource_selector.go?

**Pull request merged**